### PR TITLE
[BUG] - Missing possibility to set heading hierarchy level for some components (Issue/2849)

### LINKED ISSUE #2849: [BUG] - Missing possibility to set heading hierarchy level for some components
### Issue Description

E.g. the `Inline Notification` (other components are affected too) has a default heading level of `h5` which can't be changed by the user. This may lead to a broken heading hierarchy which can lead to decent accessibility issues and violates against "best practices" of axe-core rules engine.

### Contact Details

Slack: https://porschedev.slack.com/archives/CT7AVHTTQ/p1698063254540829

### JS Framework

all

### Porsche Design System version

3.7.0

### Expected behavior

- [ ] The user has the possibility to change default heading hierarchy level (for all affected components) to match with custom page specific hierarchy.

### Notes
- Use Component Meta to find affected components, https://designsystem.porsche.com/v3/components/introduction
- Use a standardised prop name across affected components, e.g. `headingTag: h1 | h2 | h3 | h4 | h5 | h6`
- Any component supporting the prop `heading`:
  - shall provide a prop `headingTag: h1 | h2 | h3 | h4 | h5 | h6`
  - shall provide a named slot `slot="heading"` which replaces the DOM hierarchy node within Shadow DOM
- generic `heading`, `headingTag` and `slot="heading"` behaviour and presence is ensured by test automation and/or functional component (if useful) so that components developed in the future will follow the same pattern
- Slot validation is implemented which checks for child heading nodes, e.g. `<h3 slot="heading">Some heading</h3>`
- Cross-check all affected examples in Storefront

### PR DESCRIPTION
**Pull request by o-hook**
### Pull Request Checklist

- [x] I have read and accept the [Contributor License Agreement](https://opensource.porsche.com/docs/cla)

#### References

- Preview: https://designsystem.porsche.com/issue/2849/

#### Scope

Short description of the pull request scope

#### Resolved Issue

Resolves #2849

### CODE REVIEW

File: packages/components-js/tests/e2e/specs/accordion.e2e.ts [hunk 1]
```diff
@@ -58,6 +58,8 @@ const getBody = (page: Page) => page.$('body');
 const getCollapseVisibility = async (page: Page) => getElementStyle(await getCollapsible(page), 'visibility');
 const getCollapseGridTemplateRows = async (page: Page) =>
   getElementStyle(await getCollapsible(page), 'gridTemplateRows');
+const getHeadingTagName = async (page: Page): Promise<string> =>
```
Comment (henrijoss) [inline packages/components-js/tests/e2e/specs/accordion.e2e.ts:61]: Should be able to pierce through shadow dom boundaries with `$` selector. Also Playwright encourages to use [Locators](https://playwright.dev/docs/locators).

File: packages/components-js/tests/e2e/specs/banner.e2e.ts [hunk 2]
```diff
@@ -12,18 +12,24 @@ import {
...
 const initBanner = (page: Page, opts: InitOptions): Promise<void> => {
-  const { open = false, state, dismissButton = true } = opts || {};
+  const { open = false, state, dismissButton = true, headingTag } = opts || {};
 
-  const attrs = [`open="${open}"`, state ? `state="${state}"` : '', `dismiss-button="${dismissButton}"`].join(' ');
+  const attrs = [
+    `open="${open}"`,
+    state ? `state="${state}"` : '',
+    `dismiss-button="${dismissButton}"`,
+    headingTag && `heading-tag="${headingTag}"`,
```
Comment (henrijoss) [inline packages/components-js/tests/e2e/specs/banner.e2e.ts:31]: Doesn't this produce undefined as prop on the `<p-banner>` when headingTag isn't used?

File: packages/components/src/components/inline-notification/inline-notification-styles.ts [hunk 3]
```diff
@@ -14,8 +14,22 @@ import {
   getNotificationIconJssStyle,
   getNotificationRootJssStyle,
 } from './inline-notification-styles-shared';
+import { JssStyle } from 'jss';
+import { getTypographySlottedJssStyle } from '../../styles/typography-styles';
 
 const mediaQueryMaxS = getMediaQueryMax('s');
+const textStyles = (theme: Theme): JssStyle => ({
```
Comment (henrijoss) [inline packages/components/src/components/inline-notification/inline-notification-styles.ts:21]: I think we use the `get...JssStyle` convention when naming functions which return `JssStyle`. Same for `headingStyles` as well

File: packages/components/src/index.html
Comment (henrijoss) [inline packages/components/src/index.html:1]: Revert?

File: packages/components/CHANGELOG.md [hunk 4]
```diff
@@ -14,6 +14,15 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0),
...
 ### [Unreleased]
 
+#### Added
+
+- `Inline Notification`, `Banner`: heading hierarchy can now be customized with `headingTag` prop
+  ([#3168](https://github.com/porsche-design-system/porsche-design-system/pull/3168))
+
+#### Changed
+
+- `Accordion`: `tag` property is deprecated. Use `headingTag` property instead to specify heading hierarchy level.
```
Comment (stephanschroeter) [inline packages/components/CHANGELOG.md:24]: A "diff" like we've used in the past for deprecation announements might be helpful.

File: packages/components/src/components/display/display-styles.ts [hunk 5]
```diff
@@ -33,8 +33,8 @@ export const getComponentCss = (
           ...hostHiddenStyles,
         }),
       },
-      '::slotted': {
-        [DISPLAY_TAGS.map((i) => `&(${i})`).join()]: addImportantToEachRule(getTypographySlottedJssStyle()),
+      [`::slotted(:is(${DISPLAY_TAGS.map((i) => `${i}`).join()}))`]: {
```
Comment (stephanschroeter) [inline packages/components/src/components/display/display-styles.ts:36]: Wouldn't `DISPLAY_TAGS.join()` be good enough after your changes?

File: packages/components/src/components/heading/heading-styles.ts [hunk 6]
```diff
@@ -36,8 +36,8 @@ export const getComponentCss = (
           ...hostHiddenStyles,
         }),
       },
-      '::slotted': {
-        [HEADING_TAGS.map((i) => `&(${i})`).join()]: addImportantToEachRule(getTypographySlottedJssStyle()),
+      [`::slotted(:is(${HEADING_TAGS.map((i) => `${i}`).join()}))`]: {
```
Comment (stephanschroeter) [inline packages/components/src/components/heading/heading-styles.ts:39]: same, mapping might not be needed anymore.

File: packages/components/src/components/heading/heading-styles.ts [hunk 7]
```diff
@@ -36,8 +36,8 @@ export const getComponentCss = (
...
+        ...addImportantToEachRule(getTypographySlottedJssStyle()),
```
Comment (stephanschroeter) [inline packages/components/src/components/heading/heading-styles.ts:40]: Unnecessary spread

File: packages/components/src/components/inline-notification/inline-notification-styles.ts [hunk 8]
```diff
@@ -33,16 +47,20 @@ export const getComponentCss = (
...
-        }),
+      [`::slotted(:is(${HEADING_TAGS.map((i) => `${i}`).join()}))`]: {
+        ...addImportantToEachRule(getTypographySlottedJssStyle()),
+      },
+      'slot[name="heading"]': {
+        ...headingStyles(theme),
       },
     },
+    heading: {
+      ...headingStyles(theme),
```
Comment (stephanschroeter) [inline packages/components/src/components/inline-notification/inline-notification-styles.ts:58]: Unnecessary spread

File: packages/components/src/components/inline-notification/inline-notification-styles.ts [hunk 9]
```diff
@@ -33,16 +47,20 @@ export const getComponentCss = (
...
       },
-      h5: headingSmallStyle,
-      p: textSmallStyle,
-      'h5,p': {
-        margin: 0,
-        color: getThemedColors(theme).primaryColor,
-        ...prefersColorSchemeDarkMediaQuery(theme, {
-          color: getThemedColors('dark').primaryColor,
-        }),
+      [`::slotted(:is(${HEADING_TAGS.map((i) => `${i}`).join()}))`]: {
```
Comment (stephanschroeter) [inline packages/components/src/components/inline-notification/inline-notification-styles.ts:50]: Unnecessary mapping

File: packages/components/src/components/text/text-styles.ts [hunk 10]
```diff
@@ -41,8 +41,8 @@ export const getComponentCss = (
           ...hostHiddenStyles,
         }),
       },
-      '::slotted': {
-        [TEXT_TAGS.map((i) => `&(${i})`).join()]: addImportantToEachRule(getTypographySlottedJssStyle()),
+      [`::slotted(:is(${TEXT_TAGS.map((i) => `${i}`).join()}))`]: {
```
Comment (stephanschroeter) [inline packages/components/src/components/text/text-styles.ts:44]: Unnecessary mapping

File: packages/components/src/components/text/text-styles.ts [hunk 11]
```diff
@@ -41,8 +41,8 @@ export const getComponentCss = (
...
+        ...addImportantToEachRule(getTypographySlottedJssStyle()),
```
Comment (stephanschroeter) [inline packages/components/src/components/text/text-styles.ts:45]: Unnecessary spread

File: packages/components/src/index.html [hunk 12]
```diff
@@ -1,162 +1,202 @@
 <!doctype html>
 <html dir="ltr" lang="en">
```
Comment (stephanschroeter) [inline packages/components/src/index.html:2]: please revert

File: packages/components-js/tests/e2e/specs/inline-notification.e2e.ts [hunk 13]
```diff
@@ -66,6 +70,17 @@ test('should render without button when dismissButton prop false', async ({ page
   expect(el).toBeNull();
 });
 
+test('should render correct heading tag when heading-tag property is set', async ({ page }) => {
```
Comment (stephanschroeter) [inline packages/components-js/tests/e2e/specs/inline-notification.e2e.ts:73]: Wouldn't a generic test based on component-meta information be more helpful to ensure consistency in the future and to reduce maintenance efforts?

**Comment by stephanschroeter** [comment]
Screenshot 2024-04-11 at 10 59 05
I think the second example doesn't make sense this way?

File: packages/components-js/tests/e2e/specs/accordion.e2e.ts (hunk 1 above)
Comment (o-hook) [reply packages/components-js/tests/e2e/specs/accordion.e2e.ts]: I have abstracted the test and fixed it.

File: packages/components-js/tests/e2e/specs/inline-notification.e2e.ts (hunk 13 above)
Comment (o-hook) [reply packages/components-js/tests/e2e/specs/inline-notification.e2e.ts]: Added generic test. See `components-with-heading-tag-prop.e2e.ts`

File: packages/components/src/index.html
Comment (henrijoss) [inline packages/components/src/index.html:1]: Are these changes intentional?

File: packages/components/src/components/inline-notification/inline-notification-styles.ts [hunk 14]
```diff
@@ -14,8 +14,22 @@ import {
   getNotificationIconJssStyle,
   getNotificationRootJssStyle,
 } from './inline-notification-styles-shared';
+import { JssStyle } from 'jss';
+import { getTypographySlottedJssStyle } from '../../styles/typography-styles';
 
 const mediaQueryMaxS = getMediaQueryMax('s');
+const getTextStyles = (theme: Theme): JssStyle => ({
```
Comment (henrijoss) [inline packages/components/src/components/inline-notification/inline-notification-styles.ts:21]:
```suggestion
const getTextJssStyle = (theme: Theme): JssStyle => ({
```

File: packages/components/src/components/inline-notification/inline-notification-styles.ts [hunk 15]
```diff
@@ -14,8 +14,22 @@ import {
...
+  margin: 0,
+  color: getThemedColors(theme).primaryColor,
+  ...prefersColorSchemeDarkMediaQuery(theme, {
+    color: getThemedColors('dark').primaryColor,
+  }),
+});
+
+const getHeadingStyles = (theme: Theme): JssStyle => ({
```
Comment (henrijoss) [inline packages/components/src/components/inline-notification/inline-notification-styles.ts:29]:
```suggestion
const getHeadingJssStyle = (theme: Theme): JssStyle => ({
```

File: packages/components/src/index.html
Comment (o-hook) [reply packages/components/src/index.html:1]: @henrijoss I really don't know why the diff looks like this though I haven't changed anything besides removing the contents between `<main>`.

Comment (henrijoss) [reply packages/components/src/index.html:1]: Maybe formatting? Should be able to just revert the commit to the old version :)

**Review by stephanschroeter — APPROVED** [review]
